consensus/core/src/dag_state.rs

```diff
@@ -257,6 +261,86 @@ impl DagState {
         Ok(blocks)
     }
 
+    /// Retrieves the last block proposed for the specified `authority`. If no block is found in cache
+    /// then the genesis block is returned as no other block has been received from that authority.
+    pub(crate) fn get_last_block_for_authority(&self, authority: AuthorityIndex) -> VerifiedBlock {
+        if let Some(last) = self.cached_refs[authority].last() {
+            return self
+                .recent_blocks
+                .get(last)
+                .expect("Block should be found in recent blocks")
+                .clone();
+        }
+
+        // if none exists, then fallback to genesis
+        let (_, genesis_block) = self
+            .genesis
+            .iter()
+            .find(|(block_ref, _)| block_ref.author == authority)
+            .expect("Genesis should be found for authority {authority_index}");
+        genesis_block.clone()
+    }
+
+    /// Returns the last block proposed per authority with round <= `before_round`. If `before_round`
+    /// is not provided then block of the highest available round per authority will be returned. The method will look
+    /// into cache first, otherwise will fall back in store. The method guarantees to return a result for
+    /// each authority, even if that's genesis block. In case of equivocation for an authority's last slot
+    /// only one block will be returned (the last in order).
+    pub(crate) fn get_last_block_per_authority(
+        &self,
+        before_round: Option<Round>,
+    ) -> Vec<VerifiedBlock> {
+        let before_round = before_round.unwrap_or(u32::MAX - 1);
```

Comment:
This is guaranteed to work when highest accepted round - before_round < CACHED_ROUND. But if before_round is smaller, the last block found in cache may not be the actual highest block from the authority.